fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,822 @@
+        )
+
+        # Not checked in __setup, because partial_fit may not require it.
+        if self.epochs == -1 and self.max_iter == -1:
+            raise ValueError(
+                _KWARG_ERROR_MESSAGE.format(
+                    "epochs",
+                    (
+                        "a positive number, or key-word argument "
+                        + "max_iter should be a positive number."
+                    ),
+                )
+            )
+
+        if self.batch_size == -1:
+            batch_size = X.shape[0]
+        else:
+            batch_size = self.batch_size
+        batches = ceil(X.shape[0] / batch_size)
+
+        if self.epochs == -1:
+            epochs = ceil(self.max_iter / batches)
+        else:
+            epochs = self.epochs
+
+        start_time = time()
+        last_update_time = start_time
+
+        # logger = logging.getLogger(__name__) FIXME: use logger
+
+        predictor_losses = [None]
+        adversary_losses = []
+
+        self.step_ = 0
+        for epoch in range(epochs):
+            for batch in range(batches):
+                if self.progress_updates:
+                    if (time() - last_update_time) > self.progress_updates:
+                        last_update_time = time()
+                        progress = (epoch / epochs) + (
+                            batch / (batches * epochs)
+                        )
+                        print(
+                            _PROGRESS_UPDATE.format(
+                                "=" * round(20 * progress),
+                                " " * round(20 * (1 - progress)),
+                                epoch + 1,
+                                epochs,
+                                " " * (len(str(batch + 1)) - len(str(batches))),
+                                batch + 1,
+                                batches,
+                                ((last_update_time - start_time) / progress)
+                                * (1 - progress),
+                                predictor_losses[-1],
+                                adversary_losses[-1],
+                            ),
+                            end="\n",
+                        )
+                batch_slice = slice(
+                    batch * batch_size,
+                    min((batch + 1) * batch_size, X.shape[0]),
+                )
+                (LP, LA) = self.backendEngine_.train_step(
+                    X[batch_slice], Y[batch_slice], Z[batch_slice]
+                )
+                predictor_losses.append(LP)
+                adversary_losses.append(LA)
+
+                self.step_ += 1
+
+                # Purposefully first stop and then handle callbacks
+                if self.max_iter != -1 and self.step_ >= self.max_iter:
+                    return self
+
+                if self.callbacks:
+                    stop = False
+                    if isinstance(self.callbacks, list):
```

Comment:
Small thought, rather than have this logic here, why not 'normalise' the callbacks so that even if the use passes a single callback function, `self.callbacks` will always be a list?